### DISCUSSION TITLE
Fast export to PNG on current frame button

### DISCUSSION THREAD

**Discussion by LuiToon** [opentoonz/opentoonz#6552] [Ideas]
So we can get higher definition of the photo than screenshots. One click would be more convenient than setting it up on the output settings and clicking View > Camera Box to disable the red x in the middle of the screen . I also see OpenToonz as a full-blown digital art software already so this would definitely be useful for every artists to export quickly. Thank you for reading and developing this software.

**Comment by DarrenTAnims** [comment]
The simple way is to render your images. Open the output settings, enter the frame numbers you want to render (From & To), then choose the file type (png. jpg, etc), enter a name for the image and press Render. Once this is setup for a scene, you can make further edits, then just press render (or the keyboard shortcut), so it is a one click function.

**Comment by RodneyBaker** [comment]
I add this mostly because there hasn't been any feedback yet on suggestions presented.
As such I'm adding a few more thoughts.


This request might relate to the 'Fast render to MP4' feature in that if that feature is updated to allow more options it might allow for faster render/export to PNG and more.

As @Bracket-H and @DarrenTAnims have mentioned there are currently several ways to export current frame.
@LuiToon A few clarifications will be appreciated with regard to what specifically you would see as success in exporting PNG.
When many of us think of 'Fast render' or 'Fast export' we likely think of Preview window or the 'Fast render to desktop (MP4)' option.


Some additional thoughts on what success might look like:
Thinking big this might mean that if 'Fast render to MP4' was launched while holding down shift key a dialogue box would open allowing for additional options to be set by user.

Also:  You mention 'higher resolution' and currently we don't have a lot of options for up-rezing images from their original resolution.  I note that in Output Settings we can down scale the resolution via the Shrink option but a useful addition there might be to allow negative numbers that would actually upscale to higher resolution.  I mention this specifically because of late various approaches to increasing resolution have been popular and the code for such might be readily available.  Some of these approaches reportedly use the buzz word AI and some care would need to be made to make sure the methodology used to increase resolution actually works and works well.  ;)

So to the current specs of this feature request:

Requirements
- [ ] Higher definition of the photo than screenshots
- [ ] One click process for more convenience
- [ ] Ease of setup preferrably without adjusting output settings
- [ ] Automatically turn off/ignore overlays such as Camera Box (ex:  the red x in the middle of the screen)

You might want to investigate the power of Opentoonz's Flipbook as that meets most of this criteria and has additional options.

**Comment by RodneyBaker** [comment]
Closing this as this one click process does appear to be present in Opentoonz.
As stated in commentary this does take setting up the output options (in this case to PNG) in Output Setting first in order to tell 'Render' what to render with that one click option.  Additional use of shortcut key can also make this a immediate process.

This report will be transferred to the Opentoonz docs repo for further consideration of how thorough the Render process is covered in official documentation.  There the discussion can continue to capture any shortfalls the current process still lacks.

**Comment by RodneyBaker** [comment]
I will add also that the Render option can be added to various location in Opentoonz such as xsheet toolbar and custom panels.
What that render will output of course will rely upon what is set in Output Settings as options.

Consideration should be made about how we might be able to programmatically target presets for this purpose but it would seem a set of additional and deeper shortcuts would likely need to be created to gain that fidelity in render customization for leverating in various places throughout the interface.

**Comment by RodneyBaker** [comment]
Adding here that I believe the way forward for development of the 'Fast Render to MP4' very well could be a means to customize the feature to where the user can modify and target what the Fast Render outputs.  This could be MP4 but it also could be any number of other output formats.

This suggests to me that 'Fast Render' may be the ideal place to place a customized FFMPEG output manager.
This could just output an MP4 video as it does now but it could also output and run many other processes and programs.

Some time back I created a proof of concept for this in python:

![image](https://github.com/user-attachments/assets/17369f9e-15df-4874-acab-282b73f48bdb)


What gets run by Fast Render would then depend on what is in the .ini script.
An example/test script:

```[Paths]
ffmpeg_path = C:/ffmpeg/bin/ffmpeg.exe
ffplay_path = C:/ffmpeg/bin/ffplay.exe

[Settings]
row_count = 27

[Row_1]
active = 0
index = 10
description = Play a MP4 video
command = ffplay "F:\PinDistorted.mp4"
notes = 
reference = F:\PinDistorted.mp4

[Row_2]
active = 0
index = 7
description = Loop a GIF
command = ffplay -loop 2 "F:\splash2.gif"
	
notes = 
reference = F:\splash2.gif

[Row_3]
active = 0
index = 10
description = FFMPEG GPT Link
command = https://chatgpt.com/g/g-RbR2klgoW-ffmpeggpt
notes = Click icon for AI with FFMPEG
reference = https://chatgpt.com/g/g-RbR2klgoW-ffmpeggpt

[Row_4]
active = 0
index = 4
description = Google Search
command = https://www.google.com/search?q=Opentoonz&oq=Opentoonz&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIGCAEQRRg8MgYIAhBFGDwyBggDEEUYOzIGCAQQRRg7MgYIBRBFGDsyBggGEEUYQTIGCAcQRRg80gEJNDM3M2owajE1qAIIsAIB&sourceid=chrome&ie=UTF-8
notes = Searching for 'Opentoonz'
reference = https://www.google.com/search?q=Opentoonz&oq=Opentoonz&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIGCAEQRRg8MgYIAhBFGDwyBggDEEUYOzIGCAQQRRg7MgYIBRBFGDsyBggGEEUYQTIGCAcQRRg80gEJNDM3M2owajE1qAIIsAIB&sourceid=chrome&ie=UTF-8

[Row_15]
active = 0
index = 15
description = Animation:Master Forum
command = 
notes = 
reference = https://forums.hash.com/

[Row_5]
active = 0
index = 5
description = 
command = 
notes = 
reference = 

[Row_6]
active = 0
index = 6
description = Requirements
command = 
notes = 
reference = 

[Row_7]
active = 0
index = 7
description = TODO:
command = 
notes = 
reference = 

[Row_8]
active = 0
index = 8
description = Convert Formats
command = 
notes = 
reference = 

[Row_9]
active = 0
index = 9
description = Resize images
command = ffmpeg -i splash.gif -vf "scale=iw:-1*0.33" resizedsplash.gif
notes = 
reference = 

[Row_10]
active = 0
index = 10
description = Convert Audio
command = 
notes = 
reference = 

[Row_11]
active = 0
index = 11
description = Run Opentoonz Script
command = 
notes = 
reference = 

[Row_12]
active = 0
index = 12
description = 
command = 
notes = 
reference = 

[Row_13]
active = 0
index = 13
description = 
command = 
notes = 
reference = 

[Row_14]
active = 0
index = 14
description = OT Manual (online)
command = https://opentoonz.readthedocs.io/_/downloads/en/latest/pdf/
notes = < PDF version  |  Online HTML >
reference = https://opentoonz.readthedocs.io/en/latest/

[Row_16]
active = 1
index = 1
description = 
command = "F:\runner.bat" 1
notes = 
reference = 

[Row_17]
active = 1
index = 4
description = 
command = "F:\runner.bat" 2
notes = 
reference = 

[Row_18]
active = 1
index = 2
description = 
command = "F:\runner.bat" 3
notes = 
reference = 

[Row_19]
active = 1
index = 3
description = 
command = "F:\runner.bat" 4
notes = 
reference = 

[Row_20]
active = 1
index = 5
description = 
command = "F:\runner.bat" 5
notes = 
reference = 

[Row_21]
active = 0
index = 21
description = 
command = 
notes = 
reference = 

[Row_22]
active = 0
index = 22
description = Crude Drawing program
command = F:\imagedraw.py
notes = TODO:  Won't yet run from FFM
reference = F:\imagedraw.py

[Row_23]
active = 0
index = 23
description = 
command = 
notes = 
reference = 

[Row_24]
active = 0
index = 24
description = Render a Scene (via Batch File)
command = "F:\render.bat"
notes = 
reference = F:\render.bat

[Row_25]
active = 0
index = 25
description = Render directly from Task
command = "C:\Program Files\Opentoonz\tcomposer.exe"  "F:\CustomBrushFollower\scenes\renderme.tnz" -o "F:\shrink5..png" -range 1 24 -step 1 -shrink 5 -multimedia 0 -nthreads all -maxtilesize none
notes = 
reference = 

[Row_26]
active = 0
index = 26
description = 
command = 
notes = 
reference = 

[Row_27]
active = 0
index = 27
description = 
command = 
notes = 
reference = 

```
This would not only allow Fast Export to PNG but pretty much Fast Export to any format ffmpeg (and technically speaking other programs) can output.

The specific table of commands is so that the priority of what gets output as well as ability to turn line items on/off can be stored and adjusted by the user.

**Comment by RodneyBaker** [comment]
Note here that the example ini file above was to test various possibilities.
For instance, having FFPLAY open and play back what is rendered in previous steps.
Keeping the process cross platform compatible would be important but where it comes to running other scripts/processes that would be handed over to the platform itself.

Added:  Short of this fuller implementation of a customizable command runner Fast Render is still ripe for having the ability to customize at least what gets output via Fast Render.
